src/ray/core_worker/core_worker.cc

```diff
@@ -704,6 +704,35 @@ void CoreWorker::Disconnect(
   }
 }
 
+void CoreWorker::KillLeakedProcs() {
```

Comment:
I call this in the CoreWorker. There are cases where we will still leak processes, like if the worker process segfaults. To handle this, we should keep track of child processes of the workers in the raylet. I am intentionally not doing that now to 1) get validation on this approach and 2) get a faster fix in because that kind of resource tracking is a larger change.

Note that even if we track child processes in the raylet, we'll still need code here in case the raylet dies.

Reply:
I think this is one of those cases where you do want an e2e correct solution on the first try. I don't think it makes sense to invest effort on a solution that isn't fully correct to this extent..

Reply:
What was the result of the investigation around the raylet killing the reparented child processes as the process group leader?

Reply:
> What was the result of the investigation around the raylet killing the reparented child processes as the process group leader?

Killing the reparented child processes by process group was not part of the investigation because they are in the raylet's process group (so all workers and any child processes would also get killed).

As for killing the reparented child processes, I assumed that they would be re-parented to the raylet. This was incorrect; I tried it and they are reparented to init (pid 1).

> I think this is one of those cases where you do want an e2e correct solution on the first try. I don't think it makes sense to invest effort on a solution that isn't fully correct to this extent..

The spectrum of solutions I see are as follows:

1. Kill processes by cgroup (best userspace solution).
  * Sureproof cleaning up of accidentally-leaked processes.
  * Significant re-architecture of how Ray creates/manages worker processes.
  * Requires handling Raylet crashes.

2. Kill processes by process-group.
  * Sureproof cleaning up of accidentally-leaked processes.
  * Nontrivial changes in Ray given that Ray currently places all workers in the same process group and uses that for cleanup.
  * Requires handling Raylet crashes.

3. Track worker child processes in the raylet; kill child processes once worker dies.
  * Not sureproof clean up -- will leak processes in certain race conditions, and can kill arbitrary new processes in other race conditions.
  * Straightforward but nontrivial addition to raylet.
  * Requires handling Raylet crashes.

4. Kill worker child process when a worker exits, either from Cancellation (ctrl+C) or on normal teardown.
  * Sureproof clean up of accidentally-leaked processes except in worker segfault.
  * Straightforward and small addition to core worker.
  * The raylet can crash and workers still clean up correctly.

Horable mention: Raylet kills processes that are in the Raylet process group but have ppid 1. It is not guaranteed that ppid==1 for orphaned processes.

I see (2) as the best medium-term solution. I think (4) is the better short-term solution between (3, 4) because it is sureproof under user cancellation or normal teardown, and easy to implement.

Reply:
> Horable mention: Raylet kills processes that are in the Raylet process group but have ppid 1. It is not guaranteed that ppid==1 for orphaned processes.

Why isn't this a candidate solution? It seems to (1) be correct e2e, and (2) pretty straightforward to implement via polling, for example. Even if ppid != 1, we could do it for any unknown ppid.

Reply:
Btw, https://unix.stackexchange.com/questions/149319/new-parent-process-when-the-parent-process-dies

suggests that you can use a `prctl` call to request orphaned children to be reparented to yourself.

Reply:
Oh, didn't know that we have this. Seems that's what we want.

Reply:
> >Horable mention: Raylet kills processes that are in the Raylet process group but have ppid 1. It is not guaranteed that ppid==1 for orphaned processes.
> 
> Why isn't this a candidate solution? It seems to (1) be correct e2e, and (2) pretty straightforward to implement via polling, for example. Even if ppid != 1, we could do it for any unknown ppid.

I avoided that solution because you have to distinguish between leaked child processes and non-leaked child processes (since at depth>1 child processes will have unknown ppids). And climbing the ppid tree has race conditions which are especially risky if there are many child processes being created.

That said, we could do something like a mark-and-sweep approach to drastically reduce the chance of these race conditions; on the first poll, mark candidate processes for termination, on the second poll confirm the tree structure is the same and then kill the child process.

This still fails when the raylet crashes, since the entire old raylet process group will not be known.

I think this is potentially something we do before (2) above. But even this case will require the core-worker to kill child processes on exit to handle raylet crashes (which is the code in this PR).

Reply:
>Btw, https://unix.stackexchange.com/questions/149319/new-parent-process-when-the-parent-process-dies
>
> suggests that you can use a prctl call to request orphaned children to be reparented to yourself.

Oh, nice find. Seems that we can use this so that all leaked processes get reparented to the raylet. The missing piece then would be to periodically poll children of the raylet and if a process is not recognized (i.e. isn't a known worker), then we assume it is a leaked child process and we kill it.

Can we still merge this coreworker change to handle the raylet crash case, and since we believe it addresses the high-frequency torch dataloader process leak?

Reply:
I think if raylet and worker crashes the same time, we can still have leaks.

This is my current understanding (might be wrong): with the current linux mechanisms, unless the child process is cooperative (e.g. using PR_SET_PDEATHSIG, listening to pipe), there is always corner case where the child process is leaked.